### PR TITLE
feat: traceback-to-graph correlation with ranked root causes over MCP

### PR DESCRIPTION
**Pull request by vitali87**
Implements the v1 scope of #227: a query layer that correlates a CPython traceback with the code graph and walks it backwards to rank root-cause candidates, served as two MCP tools on the existing server.

## What

- `codebase_rag/crash_correlation.py`: the engine, following `flow_verdict.py`'s discipline exactly (pure functions over `fetch_all`, client-side reachability over linear scans).
  - `parse_python_traceback` parses real CPython traceback text: the last section of a chained traceback (the failure that propagated), frame lines, indented source snippets and caret markers skipped, and the flush-left exception line (`ValueError: boom` or a bare `KeyboardInterrupt`).
  - `explain_traceback` resolves every frame through the existing dynamic-trace `FrameResolver` (same node shapes as trace ingestion, via the now-public `load_callables`) and attaches each resolved node's callers, callees, and FLOWS_TO sources; unresolved frames carry the resolver's reason (outside repo, unknown path, no match).
  - `rank_root_causes` scores candidates by three additive signals: FLOWS_TO source into the failing frame (0.6, the failing value's possible producers), presence on the crashing stack (0.3, the ground-truth path), and reverse-CALLS proximity (0.4/depth, ≤3). Each candidate carries file, definition line, reasons, and a shortest call path to the failure. Deterministic tie-breaks.
- CALLS pairs come from a dedicated fetch that keeps runtime-discovered (`static_missed`) edges: an edge a trace ingestion created is a real observed call path, exactly the evidence a root-cause walk wants.
- Graceful degradation per the issue's acceptance criteria: with no FLOWS_TO edges the ranking is CALLS-only, `flow_used` is false, and the files outside flow coverage are named so the absence is not read as evidence.
- MCP surface: `explain_traceback` and `rank_root_causes` tools, project-scoped through the server's repository root like `flow_verdict`, guarded by the same ingestor lock.

## Tests

The parser is tested against a genuine traceback produced by raising and formatting a chained exception in-test (so the shape can never drift from CPython's). The correlation and ranking tests run a planted-defect scenario: `load_config` returns None which FLOWS_TO the failing read in `handle`; it is neither on the stack nor a caller, so only the flow signal can surface it, and it ranks second of three (acceptance: top 3). Also covered: out-of-repo frames with reasons, CALLS-only degradation with coverage gaps, an unresolvable traceback, MCP registration, and both handlers end-to-end.

## Out of scope (per the issue's phasing)

The exception taxonomy, value predicates, recency ranking (phase 1 extensions), and the flight recorder (phase 2) layer on top of this surface without changing it.

Closes nothing yet; #227's v1 checklist is satisfied for Python. Compiled-language traceback parsing (the "prove the shape generalises" item) can follow as a parser-only addition since resolution and ranking are language-agnostic once frames exist.

## Summary by CodeRabbit

- **New Features**
  - Added tools to analyze Python tracebacks and identify related project code.
  - Traceback explanations show resolved frames, callers, callees, data-flow sources, and unresolved locations.
  - Added root-cause ranking with confidence scores, supporting reasons, call paths, and flow-coverage gaps.
  - Results are available in structured JSON for integrations, including chained exceptions and exception groups.

- **Documentation**
  - Added MCP server guidance for traceback analysis and root-cause ranking tools.

- **Tests**
  - Added coverage for traceback parsing, frame resolution, ranking, project boundaries, and edge cases.

### CODE REVIEW

File: codebase_rag/crash_correlation.py [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+        self.flow_gaps: tuple[str, ...] = ()
+        if not self.flow_sources:
+            gap_rows = fetch_all(CYPHER_FLOW_COVERAGE_GAPS, params)
+            self.flow_gaps = tuple(
+                sorted(
+                    path
+                    for row in gap_rows
+                    if isinstance(path := row.get(cs.KEY_PATH), str)
+                )
+            )
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/crash_correlation.py]: Fixed in 5de36bae: coverage gaps are fetched unconditionally, so a flow edge elsewhere in the project can no longer hide that the failing file sits outside coverage. Regression test: an unrelated flow edge plus an uncovered failing file now yields flow_used=true with the gap disclosed.

File: codebase_rag/crash_correlation.py [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+        current, depth = queue.popleft()
+        if depth == _CALLER_DEPTH_LIMIT:
+            continue
+        for caller in graph.callers.get(current, ()):
+            if caller in seen:
+                continue
+            seen.add(caller)
+            _, tail = reached.get(current, (0, (failing,)))
+            reached[caller] = (depth + 1, (caller, *tail))
+            queue.append((caller, depth + 1))
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/crash_correlation.py]: Fixed in 5de36bae: the reverse BFS expands callers in sorted order, so the recorded shortest path is deterministic regardless of graph row order. Diamond-path regression test added (main reaches the failure through branch_a and branch_b; branch_a's path is always the one recorded).

File: codebase_rag/crash_correlation.py [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+            if isinstance(from_qn, str) and isinstance(to_qn, str):
+                self.callers.setdefault(to_qn, []).append(from_qn)
+                self.callees.setdefault(from_qn, []).append(to_qn)
+        self.flow_sources: dict[str, list[str]] = {}
+        for row in fetch_all(CYPHER_FLOW_EDGES, params):
+            source, target = row.get("source"), row.get("target")
+            if isinstance(source, str) and isinstance(target, str):
+                self.flow_sources.setdefault(target, []).append(source)
+        self.flow_gaps: tuple[str, ...] = ()
+        if not self.flow_sources:
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/crash_correlation.py]: Fixed in 5de36bae together with the CodeRabbit thread on the same line: the coverage-gap query runs unconditionally, and flow_gaps always names uncovered files even when unrelated FLOWS_TO edges exist. Covered by a dedicated regression test.

File: codebase_rag/crash_correlation.py [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+    scores: dict[str, float] = {}
+    reasons: dict[str, list[str]] = {}
+    paths: dict[str, tuple[str, ...]] = {}
+
+    for qn, (depth, call_path) in reached.items():
+        scores[qn] = scores.get(qn, 0.0) + _SCORE_CALLER_BASE / depth
+        reasons.setdefault(qn, []).append(_REASON_CALLER.format(depth=depth))
+        paths[qn] = call_path
+    for distance, qn in enumerate(reversed(stack_qns[:-1]), start=1):
+        scores[qn] = scores.get(qn, 0.0) + _SCORE_ON_STACK
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/crash_correlation.py:327]: Addressed in 5de36bae with a disclosure rather than the suggested only-rank-when-the-last-frame-resolves gate, and here is the reasoning: the most common real-world shape is a crash propagating from inside a library (the innermost frames are site-packages and can never resolve against the project graph); refusing to rank there would discard exactly the case the tool exists for, and the graph analysis relative to the deepest project frame is still sound. What was genuinely wrong was the silent misattribution, so the report now carries anchor_is_crash_site: false whenever any traceback frame sits below the anchor (library frames and in-repo frames the graph cannot match alike), the tool description spells out that the ranking reads as relative to the deepest resolvable frame, and both cases are regression-tested. If you see a failure mode the disclosure does not cover, happy to revisit.

File: codebase_rag/crash_correlation.py [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+                    path=match.group("path"),
+                    qualname=match.group("name").strip(),
+                    line=int(match.group("line")),
+                )
+            )
+        # Source snippets and caret markers are indented; the exception line
+        # is the first flush-left line after the frames.
+        elif not line[:1].isspace() and (match := _TB_EXCEPTION.match(line)):
+            exception_type = match.group("type")
+            exception_message = match.group("message") or ""
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/crash_correlation.py:151]: Fixed in 5de36bae: the parser strips the ExceptionGroup box margin (both the + group headers and the | frame/exception lines), after which the existing last-section rule picks the last sub-exception's own traceback, the deepest real cause rather than the group wrapper. The regression test raises a genuine ExceptionGroup in-test and parses traceback.format_exc() output, so the fixture can never drift from CPython's rendering.

File: codebase_rag/crash_correlation.py [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+RETURN a.qualified_name AS from_qn, b.qualified_name AS to_qn
+"""
+
+# ``  File "/app/service.py", line 14, in handle_request``
+_TB_FRAME = re.compile(
+    r'^\s*File "(?P<path>[^"]+)", line (?P<line>\d+), in (?P<name>.+)$'
+)
+_TB_HEADER = "Traceback (most recent call last):"
+# ``ValueError: boom`` or a bare ``KeyboardInterrupt``.
+_TB_EXCEPTION = re.compile(r"^(?P<type>[A-Za-z_][\w.]*)(?::\s?(?P<message>.*))?$")
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/crash_correlation.py]: Fixed in 5de36bae: the exception-name pattern now starts with a Unicode-aware identifier class, and a regression test parses a Unicode-named exception with its message.
